dvc/remote/gdrive/__init__.py

```diff
@@ -0,0 +1,567 @@
+from __future__ import unicode_literals
+
+from time import sleep
+import threading
+import functools
+import posixpath
+import os
+import logging
+
+try:
+    import google_auth_oauthlib
+    from .oauth2 import OAuth2
+except ImportError:
+    google_auth_oauthlib = None
+
+from requests import ConnectionError
+
+from dvc.scheme import Schemes
+from dvc.path.gdrive import PathGDrive
+from dvc.utils import tmp_fname, move
+from dvc.utils.compat import urlparse, makedirs
+from dvc.remote.base import RemoteBASE
+from dvc.config import Config
+from dvc.progress import progress
+
+
+logger = logging.getLogger(__name__)
+
+
+class GDriveError(Exception):
+    pass
+
+
+class GDriveResourceNotFound(GDriveError):
+    def __init__(self, path):
+        super(GDriveResourceNotFound, self).__init__(
+            "{} resource not found".format(path)
+        )
+
+
+class track_progress(object):
+    def __init__(self, progress_name, fobj):
+        self.progress_name = progress_name
+        self.fobj = fobj
+        self.file_size = os.fstat(fobj.fileno()).st_size
+
+    def read(self, size):
+        progress.update_target(
+            self.progress_name, self.fobj.tell(), self.file_size
+        )
+        return self.fobj.read(size)
+
+    def __getattr__(self, attr):
+        return getattr(self.fobj, attr)
+
+
+def only_once(func):
+    lock = threading.Lock()
+    locks = {}
+    results = {}
+
+    @functools.wraps(func)
+    def wrapped(*args, **kwargs):
+        key = (args, tuple(kwargs.items()))
+        # could do with just setdefault, but it would require
+        # create/delete a "default" Lock() object for each call, so it
+        # is better to lock a single one for a short time
+        with lock:
+            if key not in locks:
+                locks[key] = threading.Lock()
+        with locks[key]:
+            if key not in results:
+                results[key] = func(*args, **kwargs)
+        return results[key]
+
+    return wrapped
+
+
+class RemoteGDrive(RemoteBASE):
+    """Google Drive remote implementation
+
+    Example URLs:
+
+    Datasets/my-dataset inside "My Drive" folder:
+
+        gdrive://root/Datasets/my-dataset
+
+    Folder by ID (recommended):
+
+        gdrive://1r3UbnmS5B4-7YZPZmyqJuCxLVps1mASC
+
+        (get it https://drive.google.com/drive/folders/{here})
+
+    Dataset named "my-dataset" in the hidden application folder:
+
+        gdrive://dvc/my-dataset
+
+        (this one wouldn't be visible through Google Drive web UI and
```

Comment:
Sorry if I'm missing the explanation somewhere, but could you explain why one would want something like this? Could it be `made` visible later?

Reply:
DVC remote repository is a bunch of files named by MD5 hashsums, so I guess some users would not want to see this "garbage" in the google drive UI :).

It couldn't be made visible later, the only way would be to push the repository contents to the new remote again.

Reply:
Hm... actually I don't know, it could be possible to link the repository folder to the other space, I'll check it.

Reply:
No, this is not possible - `PATCH` request with `addParents` containing the folder ID from other space returns `HTTP 200`, but does nothing.